### PR TITLE
[spaceship][deliver][pilot] add support for app clips default experience metadata and review information

### LINKED ISSUE #17639: Deliver error for App containing an App Clip
### New Issue Checklist

- [ ] Updated fastlane to the latest version
- [x] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [x] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Issue Description

Ran `deliver` this morning and got the following error:

`The request cannot be fulfilled because of the state of another resource. - Submit for review errors found.
The provided entity is missing a required relationship - You must provide a value for the relationship 'appClipVersion' with this request`

It looks like the upload fails due to the app having an `App Clip` target. Is there a way to define the meta data needed for App Clips (screenshot, title, App Clip experiences, domains, ...)?

##### Command executed
```
deliver(
  reset_ratings: false,
  submit_for_review: true,
  automatic_release: false,
  phased_release: true,
  force: true, # Skip HTML report verification
  skip_screenshots: false,
  overwrite_screenshots: true,
  ignore_language_directory_validation: true,
  app_version: version_number,
  build_number: build_number,
  skip_binary_upload: true,
  skip_metadata: false,
  metadata_path: "./fastlane/metadata",
  submission_information: {
    add_id_info_uses_idfa: false
  }
)
```

### Environment

I've redacted the `fastfile` and `appfile` as I don't think it's relevant for this issue.

<details><summary>🚫 fastlane environment 🚫</summary>

### Stack

| Key                         | Value                                                          |
| --------------------------- | -------------------------------------------------------------- |
| OS                          | 11.0.1                                                         |
| Ruby                        | 2.6.3                                                          |
| Bundler?                    | false                                                          |
| Git                         | git version 2.24.3 (Apple Git-128)                             |
| Installation Source         | /usr/local/bin/fastlane                                        |
| Host                        | macOS 11.0.1 (20B29)                                           |
| Ruby Lib Dir                | /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib |
| OpenSSL Version             | LibreSSL 2.8.3                                                 |
| Is contained                | false                                                          |
| Is homebrew                 | false                                                          |
| Is installed via Fabric.app | false                                                          |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/                    |
| Xcode Version               | 12.2                                                           |

### System Locale

| Error                       |
| --------------------------- |
| No Locale with UTF8 found 🚫 |

### fastlane gems

| Gem      | Version | Update-Status      |
| -------- | ------- | ------------------ |
| fastlane | 2.164.0 | 🚫 Update available |

### Loaded fastlane plugins:

| Plugin                  | Version | Update-Status      |
| ----------------------- | ------- | ------------------ |
| fastlane-plugin-badge   | 1.4.0   | ✅ Up-To-Date       |
| fastlane-plugin-bugsnag | 1.4.2   | 🚫 Update available |

<details><summary><b>Loaded gems</b></summary>

| Gem                       | Version      |
| ------------------------- | ------------ |
| did_you_mean              | 1.3.0        |
| slack-notifier            | 2.3.2        |
| rouge                     | 2.0.7        |
| xcpretty                  | 0.3.0        |
| terminal-notifier         | 2.0.0        |
| terminal-table            | 1.8.0        |
| plist                     | 3.5.0        |
| addressable               | 2.7.0        |
| multipart-post            | 2.0.0        |
| word_wrap                 | 1.0.0        |
| colored                   | 1.2          |
| highline                  | 1.7.10       |
| commander-fastlane        | 4.4.6        |
| faraday_middleware        | 1.0.0        |
| gh_inspector              | 1.1.3        |
| rubyzip                   | 2.3.0        |
| security                  | 0.1.3        |
| xcpretty-travis-formatter | 1.0.0        |
| uber                      | 0.1.0        |
| declarative-option        | 0.1.0        |
| representable             | 3.0.4        |
| retriable                 | 3.1.2        |
| mini_mime                 | 1.0.2        |
| httpclient                | 2.8.3        |
| google-api-client         | 0.38.0       |
| nanaimo                   | 0.3.0        |
| colored2                  | 3.1.2        |
| claide                    | 1.0.3        |
| CFPropertyList            | 3.0.2        |
| atomos                    | 0.1.3        |
| xcodeproj                 | 1.19.0       |
| unicode-display_width     | 1.7.0        |
| public_suffix             | 4.0.6        |
| tty-screen                | 0.8.1        |
| tty-cursor                | 0.7.1        |
| tty-spinner               | 0.9.3        |
| babosa                    | 1.0.4        |
| excon                     | 0.78.0       |
| unf_ext                   | 0.0.7.7      |
| unf                       | 0.1.4        |
| domain_name               | 0.5.20190701 |
| http-cookie               | 1.0.3        |
| ruby2_keywords            | 0.0.2        |
| faraday                   | 1.1.0        |
| faraday-cookie_jar        | 0.0.7        |
| fastimage                 | 2.2.0        |
| json                      | 2.3.1        |
| mini_magick               | 4.10.1       |
| dotenv                    | 2.7.6        |
| bundler                   | 1.17.3       |
| naturally                 | 2.2.0        |
| simctl                    | 1.6.8        |
| jwt                       | 2.2.2        |
| declarative               | 0.0.20       |
| multi_json                | 1.15.0       |
| signet                    | 0.14.0       |
| os                        | 1.1.1        |
| memoist                   | 0.16.2       |
| googleauth                | 0.14.0       |
| rake                      | 13.0.1       |
| digest-crc                | 0.6.1        |
| google-cloud-errors       | 1.0.1        |
| google-cloud-env          | 1.4.0        |
| google-cloud-core         | 1.5.0        |
| google-cloud-storage      | 1.29.1       |
| emoji_regex               | 3.0.0        |
| aws-eventstream           | 1.1.0        |
| aws-sigv4                 | 1.2.2        |
| aws-partitions            | 1.384.0      |
| jmespath                  | 1.4.0        |
| aws-sdk-core              | 3.109.1      |
| aws-sdk-kms               | 1.39.0       |
| aws-sdk-s3                | 1.83.1       |
| forwardable               | 1.2.0        |
| logger                    | 1.3.0        |
| stringio                  | 0.0.2        |
| ipaddr                    | 1.2.2        |
| openssl                   | 2.1.2        |
| ostruct                   | 0.1.0        |
| strscan                   | 1.0.0        |
| date                      | 2.0.0        |
| fileutils                 | 1.1.0        |
| etc                       | 1.0.1        |
| io-console                | 0.4.7        |
| zlib                      | 1.0.0        |
| libxml-ruby               | 3.1.0        |
| rexml                     | 3.1.9        |
| psych                     | 3.1.0        |
| mutex_m                   | 0.1.0        |
| webrick                   | 1.4.2        |
| curb                      | 0.9.10       |
| badge                     | 0.12.0       |
| fastlane-plugin-badge     | 1.4.0        |
| xml-simple                | 1.1.5        |
| fastlane-plugin-bugsnag   | 1.4.2        |
</details>

*generated on:* **2020-11-20**
</details>

### PR DESCRIPTION
**Pull request by telip007**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I see several green `ci/circleci` builds in the "All checks have passed" section of my PR ([connect CircleCI to GitHub](https://support.circleci.com/hc/en-us/articles/360008097173-Why-aren-t-pull-requests-triggering-jobs-on-my-organization-) if not)
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
There is currently no app clips functionality within Fastlane. My team and I were using the changes from the following PR for a while now: https://github.com/fastlane/fastlane/pull/20300, however, the author no longer maintains the PR and has no interest in getting it merged, thus we decided to open a new PR.

This PR looks to implement the goals of https://github.com/fastlane/fastlane/pull/20300, but with some minor fixes to the metadata upload and additional features such as
- Specifying metadata through the filesystem
- Passing app clip review information
- Submitting beta app clip invocation

Resolves https://github.com/fastlane/fastlane/issues/17639

#### Deliver
The following app clip default experience functionality has been added to deliver:
- option to specify an action (PLAY, VIEW, ...)
- option to specify localized subtitles
- option to specify localized default experience header image
- option to specify app clip review information

Example with metadata inside deliver:
```ruby
deliver(
  app_clip_default_experience_action: "PLAY",
  app_clip_default_experience_subtitle: {
    "en-US": "this is a test from fastlane",
    "es-MX": "está es una prueba de fastlane"
  },
  app_clip_header_images_path: "./fastlane/app-clip-header-images",
  app_clip_review_information: {
    invocation_urls: ["http://localhost.com", "http://localhost.com/1"]
  }
)
```

Example with metadata through path:
```ruby
deliver(
  app_clip_default_experience_metadata_path: "./fastlane/app-clip-default-experience-metadata",
  app_clip_header_images_path: "./fastlane/app-clip-header-images",
  app_clip_review_information: {
    invocation_urls: ["http://localhost.com", "http://localhost.com/1"]
  }
)
```

The localized header images are read from the filesystem same as screenshots:
```
$ tree app-clip-header-images/
app-clip-header-images
├── en-US
│   └── random-header-image.jpg
└── es-MX
    └── another-header-image.jpg

2 directories, 2 files
```

The metadata is read as follows:
```
$ tree app-clip-default-experience-metadata/
app-clip-default-experience-metadata
├── app_clip_default_experience_action.txt
├── en-US
│   └── app_clip_default_experience_subtitle.txt
└── es-MX
    └── app_clip_default_experience_subtitle.txt


3 directories, 3 files
```

The review information can be passed through a `invocation_urls.txt` file inside `metadata/app_clip_review_information/`.
Example:
```
http://localhost.com/3, http://localhost.com/4
```

For more information, you can also look at https://github.com/fastlane/fastlane/pull/20300.

#### Pilot
The following app clip functionality has been added to pilot:
- option to specify TestFlight app clip invocations

Example:
```ruby
upload_to_testflight(
  app_clip_invocations: [{
    title: {
      "en-US": 'Try my AppClip'
    },
    url: 'http://localhost'
  }]
)
```

#### Basic Testing Setup

1. Update Gemfile to 👇 and run bundle install, bundle update fastlane, or bundle update
```ruby
gem "fastlane", :git => 'https://github.com/pace/fastlane.git', :ref => "app-clips-default-experience"
```
2. Upload a binary to ASC with an app clip target
    - you can follow [a project tutorial provided by Apple here](https://developer.apple.com/documentation/swiftui/fruta_building_a_feature-rich_app_with_swiftui)

##### Deliver
Before specifying any deliver options:

- Confirm that fastlane uses the live version's app clip metadata for the current edit version
    - It should copy over the action, subtitle, and header image of the live version
 
After specifying the app clip default experience deliver options:

- Confirm the app clip default experience metadata specified in the deliver options reflects the ASC UI after running deliver
- Confirm that fastlane enforces that the app clip header image is exactly 1800x1200 in size
- Confirm that fastlane enforces only one app clip header image per locale
- Confirm that the additional fastlane messages via UI.{message,important,...} are appropriate

##### Pilot

- Confirm that the invocations appear on TestFlight

### CODE REVIEW

**Comment by EmDee** [comment]
Hey @rogerluan, this is the follow-up PR for https://github.com/fastlane/fastlane/pull/20300, as discussed with @blesson3. Who do you think can help review this PR?

**Comment by rogerluan** [comment]
Hi @EmDee 👋  I've tagged a couple people who could help with reviewing, and I could as well, but I've been busy and haven't been able to contribute as much time to fastlane as I used to 😞 

Just as a reminder, this PR doesn't have to be reviewed+merged for you guys (or anyone) to start using it, you can just point your fastlane installation directly to this branch by modifying your Gemfile as:

```ruby
gem 'fastlane', git: 'https://github.com/pace/fastlane.git', branch: 'app-clips-default-experience'
```

And running `bundle install` to apply the changes. I hope this helps unblock you guys! 🙏 Sorry if I couldn't be more helpful this time

**Comment by EmDee** [comment]
Thanks @rogerluan! We have been using the changes in here for close to a year now, it's just a slight inconvenience to always have to keep this branch up to date with the main branch, therefore it would be nice to have this merged.

**Comment by telip007** [comment]
@ainame @revolter @rogerluan do you guys have an ETA to when this might be reviewed and hopefully merged? 🙂

**Comment by rogerluan** [comment]
No ETA from my end at this point, sorry 😞 been busy with other life priorities, I hope you understand 🙏

**Comment by telip007** [comment]
@rogerluan of course! no worry, I just wanted to ping again and check on it :)

**Review by revolter — CHANGES_REQUESTED** [review]
I don't have experience with App Clips, and I also don't have time to test this, but I did take a look over all the changes, and, except the notes that I've left, I didn't see other worrisome things.

So, since you have been using this for a year, and it mainly adds new logic, I'd merge this in order to simplify the lives of the people that need this.

@rogerluan, what do you think? 😁

**Comment by danl3v** [comment]
Definitely an advocate of this PR! I finally decided to remove my app clip since it would not work on my ci pipeline, but I can add back once this is merged

**Comment by revolter** [comment]
@telip007, do you have time to address the review notes that I've left?

**Comment by TemperMichael** [comment]
Any updates on this? @telip007 Looking forward to use that, would save me quite some time 😅

**Comment by fnuky** [comment]
Please can be this PR finished? It would help me to solve my problem

**Comment by telip007** [comment]
Sorry, I was on parental leave and did not check GitHub. Will address the notes in the course of the day.

File: deliver/lib/deliver/loader.rb
Comment (telip007) [reply deliver/lib/deliver/loader.rb]: Used the suggested change.

File: deliver/lib/deliver/runner.rb
Comment (telip007) [reply deliver/lib/deliver/runner.rb:144]: You are right, I thought I had it in place for `upload_metadata` as well. This should now be included.

File: fastlane/lib/fastlane/actions/docs/upload_to_app_store.md.erb
Comment (telip007) [reply fastlane/lib/fastlane/actions/docs/upload_to_app_store.md.erb]: Used the suggested change.

File: fastlane/lib/fastlane/actions/docs/upload_to_app_store.md.erb
Comment (telip007) [reply fastlane/lib/fastlane/actions/docs/upload_to_app_store.md.erb]: Used the suggested change.

File: fastlane/lib/fastlane/actions/docs/upload_to_app_store.md.erb
Comment (telip007) [reply fastlane/lib/fastlane/actions/docs/upload_to_app_store.md.erb]: Used the suggested change.

**Comment by hoppsen** [comment]
Any chance we can pick this one back up again? @telip007

**Comment by pf-AliKemal** [comment]
This fix will help me to solve my problem. It would be great if you could merge/complete this PR

**Comment by pf-mahmoud** [comment]
I have the same issue when happing AppClips, @telip007 Could you finalize this PR please?

**Comment by telip007** [comment]
Hey all, sorry for letting you wait. I'm currently running at total capacity, so finding time for this is a bit hard. I can't tell you when I will be able to continue working on this again.

**Comment by sa-ma** [comment]
Hello, any update on this?

**Comment by pf-AliKemal** [comment]
Hi @telip007,  Do you have any update on this?

**Comment by pf-mahmoud** [comment]
Hi @telip007 ,
Any update, or any timeline for handling this?

**Comment by stephengilroy-grx** [comment]
Hey @revolter @ainame 👋 ! Do you think you could review @sa-ma PR referenced above? https://github.com/fastlane/fastlane/pull/22041

I think it was based off this one and seems like it might be the best path forward to getting this functionality in fastlane?

**Comment by revolter** [comment]
Closed in favour of #22041.

**Comment by NickAtGit** [comment]
Is there a way to download the metadata from ASC @telip007? I tried `fastlane deliver download_metadata` but it's not pulling the AppClip metadata. Any idea? Thanks in advance!

**Pull request closed without merging**
